src/workflows/pdf/PdfWorkflowManager.ts

```diff
@@ -161,55 +185,152 @@ export default class PdfWorkflowManager<
-      throw new Error('`cbFinalizePdf` is not defined!');
-    }
+      if (this.cbFinalizePdf === null) {
+        throw new Error('`cbFinalizePdf` is not defined!');
+      }
 
-    for (let rootIdx = 0; rootIdx < this.data.length; rootIdx++) {
-      if (!this.factoriesMeta.has(rootIdx)) {
-        logger.logError(
-          this.logPrefix + `'${rootIdx}' is missing from 'factoriesMeta'`,
-          {
-            factoriesMeta: this.factoriesMeta,
-          }
-        );
-        throw new Error(`'${rootIdx}' is missing from 'factoriesMeta'`);
+      for (let rootIdx = 0; rootIdx < this.data.length; rootIdx++) {
+        if (!this.factoriesMeta.has(rootIdx)) {
+          logger.logError(
+            this.logPrefix + `'${rootIdx}' is missing from 'factoriesMeta'`,
+            {
+              factoriesMeta: this.factoriesMeta,
+            }
+          );
+          throw new Error(`'${rootIdx}' is missing from 'factoriesMeta'`);
+        }
+
+        const { meta, metaCountedPages } = this.factoriesMeta.get(rootIdx)!;
+        const item = this.data[rootIdx];
+
+        pageNumber = this.cbFinalizePdf({
+          rootToC,
+          filePaths,
+          meta,
+          metaCountedPages,
+          data: item,
+          pageNumber,
+        });
       }
 
-      const { meta, metaCountedPages } = this.factoriesMeta.get(rootIdx)!;
-      const item = this.data[rootIdx];
+      const mergedPdfPath = mergePDF(filePaths);
+      const pdfWithToCPath = writeToC(mergedPdfPath, rootToC);
+
+      failSafeDeleteFiles([mergedPdfPath]);
 
-      pageNumber = this.cbFinalizePdf({
-        rootToC,
-        filePaths,
-        meta,
-        metaCountedPages,
-        data: item,
-        pageNumber,
+      logger.logDebug(this.logPrefix + 'PDF merged', {
+        pdfWithToCPath,
+        ids: this.entityIds,
       });
+
+      this.taskFinished(pdfWithToCPath);
     }
+  }
 
-    const mergedPdfPath = mergePDF(filePaths);
-    const pdfWithToCPath = writeToC(mergedPdfPath, rootToC);
+  private createZipFile(fileList: Files[]) {
+    const outPath = `${generateTmpPath()}.zip`;
+    const archive = archiver('zip', { zlib: { level: 9 } });
+    const stream = createWriteStream(outPath);
+
+    return new Promise<string>(async (resolve, reject) => {
+      stream.on('close', () => {
+        logger.logDebug(
+          `${
+            this.logPrefix
+          } zip file has been finalized and is ${archive.pointer()} total bytes`,
+          {}
+        );
+      });
+      archive.on('error', (err) => {
+        return reject(err);
+      });
 
-    failSafeDeleteFiles([mergedPdfPath]);
+      fileList.forEach((file) => {
+        archive.file(file.path, {
+          name: file.name,
+        });
+      });
 
-    logger.logDebug(this.logPrefix + 'PDF merged', {
-      pdfWithToCPath,
-      ids: this.entityIds,
+      archive.pipe(stream);
+      stream.on('close', () => {
+        return resolve(outPath);
+      });
+      archive.finalize();
```

Comment:
I don't see the reason why we having two stream.on('close', ()=>{}) 
The first one seems to simply show logs? can we move the log to the second stream.on instead?
I see this same pattern occurred in /ZipWorkflowManager.ts  as well